File: python/wordladder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        """
        if endWord not in wordList:
            return 0
        que = collections.deque()
        que.append((beginWord, 1))
        #vis = set()
        while que:
            target, dis = que.popleft()
            if target == endWord:
                return dis
            tmp = []
            for s in wordList:
                if self.isValid(target, s):
                    que.append((s, dis+1))
                else:
                    tmp.append(s)
            wordList = tmp
        return 0
    
    def isValid(self, cur, next):
        cnt = 0
        for i in xrange(len(cur)):
            if cnt > 1:
                return False
            if cur[i]!=next[i]:
                cnt += 1
        return cnt == 1
        """
        length = 2
        if endWord not in wordList:
            return 0
        front, back = set([beginWord]), set([endWord])
        wordDict = set(wordList)
        wordDict.discard(beginWord)
        
        while front:
            
            front = wordDict & (set(word[:index] + ch + word[index+1:] for word in front 
                                for index in range(len(beginWord)) for ch in 'abcdefghijklmnopqrstuvwxyz'))
            
            if front & back:
                return length
            length += 1
            if len(front) > len(back):
                front, back = back, front
            wordDict -= front
        return 0
        
        

File: cpp/wordladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> set;
        for (auto &s: wordList) set.insert(s);
        if (set.find(endWord) == set.end()) return 0;
        queue<string> que;
        que.push(beginWord);
        int count = 0;
        while (!que.empty()) {
            int size = que.size();
            count++;
            for (int i = 0; i < size; i++) {
                auto target = que.front();
                que.pop();
                if (target == endWord) return count;
                for (int j = 0; j < target.size(); j++) {
                    string tmp = target;
                    for (int k = 0; k < 26; k++) {
                        target[j] = 'a' + k;
                        if (set.find(target) != set.end()) {
                            que.push(target);
                            set.erase(target);
                        }
                    }
                    target = tmp;
                }
                
            }
        }
        return 0;
    }
};
